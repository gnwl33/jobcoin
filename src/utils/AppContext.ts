import React from 'react';

export type ContextType = {
  user: string;
  setUser?: React.Dispatch<React.SetStateAction<string>>;
};

export default React.createContext<ContextType>({
  user: '',
  setUser: () => {},
});
