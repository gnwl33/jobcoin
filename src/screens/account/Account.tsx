import React, {useState, useContext} from 'react';
import {
  View,
  SafeAreaView,
  Text,
  StyleSheet,
  ScrollView,
  KeyboardAvoidingView,
  RefreshControl,
  Platform,
} from 'react-native';
import {Button as TextButton} from 'react-native-paper';
import HistoryGraph from './HistoryGraph';
import SendWidget from './SendWidget';
import {Layout as l} from '../../styles';
import AppContext from '../../utils/AppContext';
import useFetch from '../../utils/useFetch';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {UserInfo, RootStackParamList} from '../../utils/types';

export type AccountProps = NativeStackScreenProps<
  RootStackParamList,
  'Account'
>;

const Account = ({navigation}: AccountProps) => {
  const {user} = useContext(AppContext);
  const [refreshing, setRefreshing] = useState(false);

  const {data, fetchData} = useFetch<UserInfo>(
    `http://jobcoin.gemini.com/skyline-departed/api/addresses/${user}`,
  );

  const userInfo = data;
  const userBalance = userInfo?.balance;
  const balanceLength = userBalance?.length || 0;

  const onRefresh = async () => {
    setRefreshing(true);
    await fetchData();
    setRefreshing(false);
  };

  const viewPadding = Platform.OS === 'ios' ? s.ios : s.android;

  return (
    <SafeAreaView>
      <ScrollView
        refreshControl={
          <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
        }>
        <KeyboardAvoidingView style={[l.center, viewPadding]} behavior="height">
          <View style={[s.container, l.row, l.aICenter, l.wide100]}>
            <Text style={s.greeting}>
              Hi, {user.length < 12 ? user : user.slice(0, 11) + '...'}
            </Text>
            <TextButton
              style={s.signOut}
              onPress={() => {
                // setUser('');
                navigation.navigate('SignIn');
              }}
              mode="text">
              Sign out
            </TextButton>
          </View>
          <>
            <Text style={s.balanceLabel}>Jobcoin Balance</Text>
            <Text testID="balance" style={s.balanceAmount}>
              {balanceLength < 11
                ? userBalance
                : userBalance?.slice(0, 10) + '...'}
            </Text>
          </>
          <HistoryGraph userInfo={userInfo} />
          <SendWidget fetchData={fetchData} />
        </KeyboardAvoidingView>
      </ScrollView>
    </SafeAreaView>
  );
};

export default Account;

const s = StyleSheet.create({
  ios: {
    paddingBottom: 70,
  },
  android: {
    paddingTop: 10,
    paddingBottom: 20,
  },
  container: {
    marginTop: 30,
  },
  greeting: {
    marginLeft: 26,
    fontSize: 34,
  },
  signOut: {
    position: 'absolute',
    top: 5,
    right: 12,
  },
  balanceLabel: {
    fontSize: 21,
    marginTop: 35,
  },
  balanceAmount: {
    fontSize: 50,
  },
});
