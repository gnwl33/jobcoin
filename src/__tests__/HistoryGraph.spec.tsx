import 'react-native';
import React from 'react';
import Account from '../screens/account/Account';
import {render} from '@testing-library/react-native';
import AppContext from '../utils/AppContext';

describe('Account', () => {
  it('shows correct data on initial render', async () => {
    global.fetch = jest.fn().mockResolvedValueOnce({
      json: () =>
        Promise.resolve({
          balance: '20.25',
          transactions: [
            {
              timestamp: '2014-04-22T13:10:01.210Z',
              toAddress: 'BobsAddress',
              amount: '50.35',
            },
            {
              timestamp: '2014-04-23T18:25:43.511Z',
              fromAddress: 'BobsAddress',
              toAddress: 'AlicesAddress',
              amount: '30.1',
            },
          ],
        }),
    });
    const wrapper = render(
      <AppContext.Provider value={{user: 'BobsAddress'}}>
        {/* @ts-ignore */}
        <Account />
      </AppContext.Provider>,
    );

    const lineChart = await wrapper.findByTestId('lineChart');
    expect(lineChart.props).toMatchSnapshot();
    expect(lineChart.props.children[0].props.data.datasets[0].data).toEqual([
      50.35, 20.25,
    ]);
  });

  //    it('updates after sending Jobcoins', () => {
  //    });

  //    it('updates after refreshing', () => {
  //    });
});
