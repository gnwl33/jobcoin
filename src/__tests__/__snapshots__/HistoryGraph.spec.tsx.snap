// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "children": Array [
    <LineChart
      chartConfig={
        Object {
          "backgroundGradientFrom": "rgba(0,0,0,0)",
          "backgroundGradientFromOpacity": 0,
          "backgroundGradientTo": "rgba(0,0,0,0)",
          "backgroundGradientToOpacity": 0,
          "color": [Function],
          "fillShadowGradient": "#65DEFA",
          "propsForDots": Object {
            "r": "5",
          },
          "strokeWidth": 2,
        }
      }
      data={
        Object {
          "datasets": Array [
            Object {
              "color": [Function],
              "data": Array [
                50.35,
                20.25,
              ],
              "strokeWidth": 2,
            },
          ],
          "labels": Array [],
        }
      }
      formatYLabel={[Function]}
      height={250}
      onDataPointClick={[Function]}
      style={
        Object {
          "marginBottom": -30,
          "marginLeft": -7,
          "marginTop": 20,
        }
      }
      width={750}
      withHorizontalLabels={true}
      withInnerLines={false}
      withOuterLines={false}
      withShadow={false}
      withVerticalLabels={false}
    />,
    <View>
      <Text
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        20.25 - Apr 23, 2:25PM
      </Text>
    </View>,
  ],
  "testID": "lineChart",
}
`;
